### PR TITLE
Avoid generated-sources clash with MapStruct in codegen

### LINKED ISSUE #11253: Regression in 1.7: Dev mode auto-compile fails in MapStruct with "Attempt to recreate a file for type ..."
**Describe the bug**
I am getting this strange error during auto-compile while trying to start dev mode in one of our modules:
```
[INFO] --- quarkus-maven-plugin:1.7.0.CR2:dev (default-cli) @ middleware-patient-api-pro-fe ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] Copying 2 resources
[INFO] Changes detected - recompiling the module!
[INFO] Compiling 54 source files to C:\Develop\someproject\middleware\_develop-2\someproject-middleware\patient\api\pro-fe\target\classes
[INFO] -------------------------------------------------------------
[ERROR] COMPILATION ERROR :
[INFO] -------------------------------------------------------------
[ERROR] /C:/Develop/someproject/middleware/_develop-2/someproject-middleware/patient/api/pro-fe/src/main/java/com/somecompany/someproject/middleware/patient/api/pro_fe/data/mapper/CaseFileApiMapper.java:[26,8] Internal error in the mapping processor: java.lang.Rat org.codehaus.plexus.classworlds.launcat org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:347)  Caused by: javax.annotation.processing.FilerException: Attempt to recreate a file for type com.somecompany.someproject.middleware.patient.api.pro_fe... 49 more  Casat jdk.compiler/com.sun.tools.javac.processing.JavacFiler.createSourceFile(JavacFiler.java:426)  stence(at org.mapstruct.ap.internal.processor.MapperRenderingProcessor.createSourceFile(Mappe[INFO] 1 error
```
I have no clue why this is so horribly garbled. `mvn -e ...` doesn't help much but at least reveals which type was attempted to recreate.
```
Attempt to recreate a file for type com.somecompany.someproject.middleware.patient.api.pro_fe.data.mapper.CaseFileApiMapperImpl
    at org.mapstruct.ap.internal.processor.MapperRen(... garbled, I suppose MapperRenderingProcessor)
```

This is _not_ flaky, btw. Only the affected "MapperImpl" can fluctuate (we have multiple mappers).

Dev mode in another module does work, but there we do not use MapStruct.

**The very same code works with 1.6.1.Final!**

**Expected behavior**
Should just compile (if needed) without errors.

**Actual behavior**
Build is aborted.

**To Reproduce**
1. git clone clone https://github.com/famod/quarkus-quickstarts
2. git switch issue-11253-mapstruct-vs-prepare 
4. cd getting-started
5. should pass: mvn clean verify (or install)
6. should fail: mvn quarkus:dev

Note: This reproducer uses MapStruct 1.4.0.Beta3. Downgrading to 1.3.1.Final does not change anything.

**Configuration**
n/a

**Environment (please complete the following information):**
 - Output of `uname -a` or `ver`: `MINGW64_NT-10.0-18363 XXX 3.0.7-338.x86_64 2019-11-21 23:07 UTC x86_64 Msys`
 - Output of `java -version`: `OpenJDK 64-Bit Server VM AdoptOpenJDK (build 11.0.7+10, mixed mode)`
 - GraalVM version (if different from Java): n/a
 - Quarkus version or git rev: 1.7.0.CR1 or 1.7.0.CR2
 - Build tool (ie. output of `mvnw --version` or `gradlew --version`): Apache Maven 3.6.3

**Additional context**
`mvn clean compile quarkus:dev` works, which is no wonder since compile is skipped in this case.

I found two commits that may have something to do with it (the first one being the "hottest" candidate):
- https://github.com/quarkusio/quarkus/commit/93e1ba6f97fa2714fe67c42d8d197f0210437100#diff-c007939524ea2a3177991ef913abf5df
- https://github.com/quarkusio/quarkus/commit/3d7e64230f97a1772fb6bb58b929a2471e5a5eb2#diff-c007939524ea2a3177991ef913abf5df

I did not yet have the time to debug this.

### PR DESCRIPTION
**Pull request by famod**
Fixes #11253 

This avoids recreation attempts of annotation processors like MapStruct that are writing to `${project.build.directory}/generated-sources/annotations` by default.
See also: http://maven.apache.org/plugins/maven-compiler-plugin/compile-mojo.html#generatedSourcesDirectory

The additon of `${project.build.directory}/generated-sources` by `CodeGenMojo` _seems_ to make MapStruct (or any other processor) see its own output a second time. This change is preventing this by introducing a disjunct subdirectory.

### CODE REVIEW

**Comment by famod** [comment]
FTR, there is an alternate approach: #11311

**Review by gastaldi — APPROVED** [review]
LGTM

**Comment by famod** [comment]
@michalszynkiewicz I would be great if you could have a look at this.

Btw, this PR does not change the _unconditional_ execution of the prepare goal by the dev goal.
I suggest to rethink that. Dev mode might be fast to start in small projects, but it definitely takes its time in larger, multi-module projects which is why any overhead should be avoided.

**Comment by famod** [comment]
`formatter-maven-plugin` seems a bit overzealous at times...

**Comment by gastaldi** [comment]
Not directly related, but it seems that MapStruct should also generate sources in a disjunct directory, avoiding using the `generated-sources` directory directly

**Comment by famod** [comment]
@gastaldi It is implicitly using `generated-sources/annotations` (the default) which should be ok for annotation processors.

**Review by gsmet — CHANGES_REQUESTED** [review]
Let's wait for @aloubyansky and @michalszynkiewicz 's approvals before merging this one.

File: devtools/maven/src/main/java/io/quarkus/maven/CodeGenMojo.java
Comment (aloubyansky) [inline devtools/maven/src/main/java/io/quarkus/maven/CodeGenMojo.java:106]: I think we may need both in case of testing.

Comment (aloubyansky) [reply devtools/maven/src/main/java/io/quarkus/maven/CodeGenMojo.java:106]: Or pass both to sourceRegistrar.accept(generatedSourcesDir) and keep the test dir CodeGenerator.init.

Comment (famod) [reply devtools/maven/src/main/java/io/quarkus/maven/CodeGenMojo.java:106]: Test execution comes from a separate goal. So there is one goal for the "main" part and one for the test part (if there is something to generate for tests at all), each one (potentially) adding a different directory.
Or do you mean the dev goal should also trigger prepare for the test part?

Comment (aloubyansky) [reply devtools/maven/src/main/java/io/quarkus/maven/CodeGenMojo.java:106]: Ah, right, it's invoked for each goal.

**Review by aloubyansky — APPROVED** [review]

**Comment by gsmet** [comment]
@famod please don't set older versions. I do it when backporting. Until then it must target master and have the `triage/backport?` label.

**Comment by famod** [comment]
@gsmet Ok, will do. I actually had a look at the new COMMITTERS.md and was still a bit unsure after that.

**Comment by michalszynkiewicz** [comment]
I think the proper fix may be to remove the following line instead, checking it https://github.com/quarkusio/quarkus/blob/bbb79f107a3a8ac661618b9b3ba55c022d311c94/devtools/maven/src/main/java/io/quarkus/maven/CodeGenMojo.java#L109

**Comment by michalszynkiewicz** [comment]
PR for ^ https://github.com/quarkusio/quarkus/pull/11328

**Comment by famod** [comment]
FTR, I left at #11328. Summary: A dedicated subdirectory still seems like the cleanest solution, IHMO.

**Comment by famod** [comment]
With #11328 now merged, this PR has become optional. If we want to make extra sure to avoid clashes it can be merged, otherwise it should be closed.

**Comment by michalszynkiewicz** [comment]
On one hand, we are even more separate with changing `generated-sources` to sth else here, on the other, subdirectories of `generated-sources` seem to be the standard location for generated sources ;)

**Comment by michalszynkiewicz** [comment]
@aloubyansky WDYT?

**Comment by aloubyansky** [comment]
I think it makes sense. If you execute `package`, both test and app sources are generated, right? if they are in the same location, will all of them end up being packaged?

**Comment by michalszynkiewicz** [comment]
What we have now in master puts the "normal" generated sources in the `tartget/generated-sources` dir and the test ones in `target/generated-test-sources`

The test mojo adds stuff as test compile source root: https://github.com/quarkusio/quarkus/blob/master/devtools/maven/src/main/java/io/quarkus/maven/CodeGenTestMojo.java#L18

While the normal one adds it as a compile source root: https://github.com/quarkusio/quarkus/blob/master/devtools/maven/src/main/java/io/quarkus/maven/CodeGenMojo.java#L62

I don't think the test classes would sneak into user's app.

This PR shouldn't change this behavior

**Comment by aloubyansky** [comment]
Then it's your call @michalszynkiewicz

**Comment by michalszynkiewicz** [comment]
@famod I will close this PR, we can resurrect it if we have problems in the future.
Thanks a lot for your work on it, it sped up finding the unnecessary sources registration fixed in #11253

**Comment by famod** [comment]
Glad I could help!

**Pull request closed without merging**
